massa-bootstrap/src/server.rs

```diff
@@ -14,363 +42,470 @@ use massa_signature::KeyPair;
+                });
+            } else {
+                let _ = match  self.runtime.block_on(server.send_error("Bootstrap failed because the bootstrap server currently has no slots available.".to_string())) {
                             Err(_) => Err(std::io::Error::new(std::io::ErrorKind::TimedOut, "bootstrap error no available slots send timed out").into()),
                             Ok(Err(e)) => Err(e),
                             Ok(Ok(_)) => Ok(()),
                         };
-                        debug!("did not bootstrap {}: no available slots", remote_addr);
-                    }
-                }
+                debug!("did not bootstrap {}: no available slots", remote_addr);
             }
         }
-        // wait for bootstrap sessions to finish
-        while bootstrap_sessions.next().await.is_some() {}
-
+        // TODO: when they are synchrenous, consider if we want to signal updater and listener
+        //       here, or in the stop method.
+        // TODO: do we drop(self.runtime) here?
         Ok(())
     }
 
-    #[cfg(test)]
-    // TODO we didn't test whether the peer IP address is banned
-    async fn is_ip_allowed(
-        &self,
-        remote_addr: SocketAddr,
-        server: BootstrapServerBinder,
-        _whitelist: &Option<HashSet<IpAddr>>,
-        _blacklist: &Option<HashSet<IpAddr>>,
-    ) -> io::Result<(BootstrapServerBinder, SocketAddr)> {
-        Ok((server, remote_addr))
-    }
-
-    #[cfg(not(test))]
-    // whether the peer IP address is banned
-    async fn is_ip_allowed(
-        &self,
-        remote_addr: SocketAddr,
-        mut server: BootstrapServerBinder,
-        whitelist: &Option<HashSet<IpAddr>>,
-        blacklist: &Option<HashSet<IpAddr>>,
-    ) -> io::Result<(BootstrapServerBinder, SocketAddr)> {
-        let ip = normalize_ip(remote_addr.ip());
-        // whether the peer IP address is blacklisted
-        let not_allowed_msg = if let Some(ip_list) = &blacklist && ip_list.contains(&ip) {
-            massa_trace!("bootstrap.lib.run.select.accept.refuse_blacklisted", {"remote_addr": remote_addr});
-            Some(format!("IP {} is blacklisted", &ip))
-        // whether the peer IP address is not present in the whitelist
-        } else if let Some(ip_list) = &whitelist && !ip_list.contains(&ip){
-            massa_trace!("bootstrap.lib.run.select.accept.refuse_not_whitelisted", {"remote_addr": remote_addr});
-            Some(format!("A whitelist exists and the IP {} is not whitelisted", &ip))
+    /// this is a bit complicated, but the simple is bad:
+    /// ```rust-ignore
+    /// if self.stopper_rx.try_recv().is_ok() {
+    ///     break;
+    /// }
+    /// // if the server is stopped while blocked here, this risks never unblocking
+    /// // OR: would serve as an OBO error; Server doesn't actually stop working until after
+    /// // it's done processing one more connection _after_ it's been stopped.
```

Comment:
I don't understand this sentence

Reply:
does this clarify?
```rust
    /// // if the server is stopped while blocked on the next line, this risks never unblocking
    /// //
    /// // An approach that somehow forces a recv in the event that the stop signal is sent,
    /// // it would serve as an OBO error; Server doesn't actually stop working until after
    /// // it's done processing one more connection _after_ it's been stopped.
    /// //
    /// // ...any bug fix at that point would be a downstream hack to an upstream
    /// // bug.
```

Reply:
It's still confusing for me maybe I'm not the brightest light but "forces a recv in the event that the stop signal is sent" => it's not "in the channel that the stop signal is sent"?

Reply:
> I'm not the brightest light

Even if that were true, I don't think it's relevant: This explanation is almost as incoherent as a Charles Manson rant...

In any case, it's probably better to get rid of it. It's commenting as to how the code _isn't_, which reminds me of a lesson I learned in my educator training: Don't teach by demonstrating how things _shouldn't_ be done. That's just bad.

Do you feel that, with just the comments on how it _does_ work, that things are (or could be made to be) clear enough?